src/compliance.cc

```diff
@@ -0,0 +1,55 @@
+// Copyright 2023 Google LLC
+// SPDX-License-Identifier: BSD-2-Clause
+
+#if defined(AVIF_ENABLE_COMPLIANCE_WARDEN)
+
+#include <cstddef>
+#include <cstdint>
+#include <limits>
+
+#include "avif/internal.h"
+#include "box_reader_impl.h"
+#include "spec.h"
+
+bool checkComplianceStd(Box const & file, SpecDesc const * spec);
+
+SpecDesc const * specFind(const char * name);
+std::vector<SpecDesc const *> & g_allSpecs();
+extern const SpecDesc * const globalSpecAvif;
+extern const SpecDesc * const globalSpecAv1Hdr10plus;
+extern const SpecDesc * const globalSpecHeif;
+extern const SpecDesc * const globalSpecIsobmff;
+extern const SpecDesc * const globalSpecMiaf;
+
+avifResult avifIsCompliant(uint8_t * data, size_t size)
+{
+    // See compliance_warden.sh.
+    if (g_allSpecs().empty()) {
+        registerSpec(globalSpecAvif);
+        registerSpec(globalSpecAv1Hdr10plus);
+        registerSpec(globalSpecHeif);
+        registerSpec(globalSpecIsobmff);
+        registerSpec(globalSpecMiaf);
+    }
+
+    // Inspired from ComplianceWarden/src/app/cw.cpp
+    BoxReader topReader;
+    for (char sym : { 'f', 'i', 'l', 'e', '.', 'a', 'v', 'i', 'f' }) {
+        // Setting made-up file name (letter by letter).
+        topReader.myBox.syms.push_back({ "filename", static_cast<int64_t>(sym), 8 });
+    }
+    AVIF_CHECKERR(size <= std::numeric_limits<int>::max(), AVIF_RESULT_INVALID_ARGUMENT);
```

Comment:
nit: move this to the top of the function

Reply:
I usually prefer to keep the check close to the relevant line, but noop in case of failure is a common pattern too. I guess keeping the check below is ~noop anyway (besides the `registerSpec()` trick) so leaving it as is.